### PR TITLE
Eliminate usagestatsdeprecated package

### PR DESCRIPTION
**Pull request by dadlerj**
I'm not 100% sure who to tag on this—please just let me know if any of you are not the right person!

This update does three things:

* It (finally) eliminates the old `usagestatsdeprecated` package. For historical context, this was a redis-based precursor to our postgres-based internal event logging architecture. It was removed from most places a long time ago, but has remained the data source behind user counts for billing purposes (i.e., those that admins saw if they went to the Admin > Usage Statistics page, or if they queried users from the API). Why was this left in place for so long? There were concerns in the past that some customers wouldn't be using the postgres-based system, but I don't believe that's even possible anymore.

* It updates all calls to that package to use `usagestats` instead. Similar functions already existed, so this didn't require much updating.

* However, the `usagestats` package captures more events than the old `usagestatsdeprecated` one did! For example, `usagestats` records events from all users, not just registered ones. `usagestats` records backend events (e.g. the `"ping"` event which is logged every time the instance sends a ping, or events that fire when saved searches run, or events that fire when batch changes are created via API, etc. So I added a parameter to the database functions that return unique user counts to exclude system users.

One open question from my end: should I make it so that system users are excluded by default, instead of requiring the dev to set a flag?

I also identified a new bug while testing this: https://github.com/sourcegraph/sourcegraph/issues/38825

## Test plan

* Tested in local instance
* Validated that all unit tests still worked

...Any other suggestions?

## Screenshots

The new admin usage stats page on a fresh instance:

image

The corresponding list of events up to that point:

image

Note that even though there are a number of system user events, they are not showing up on the active users page.

### CODE REVIEW

**Review by unknwon — APPROVED** [review]

**Comment by unknwon** [comment]
> One open question from my end: should I make it so that system users are excluded by default, instead of requiring the dev to set a flag?

I think it's just a matter of taste, but let's just don't change the default behavior.

---

Just to be on the safe side, I think it makes sense to merge this PR after 3.42 branch cut which happens on July 20th. In case we shipped broken admin usage stats to customers and they freaked out ... then we have about a month to validate on dotcom to see if data are making sense.

**Review by philipp-spiess — APPROVED** [review]
Noice!

Pizza Guy Thumb

File: client/jetbrains/webview/src/telemetry/EventLogger.ts [hunk 1]
```diff
@@ -22,22 +16,6 @@ events
         bufferTime(1000),
         concatMap(events => {
             if (events.length > 0) {
-                // For every IDESearchSubmitted, we also fire a legacy logUserEvent event. This is
-                // necessary since we only look at the latter to calculate some activity usage
-                // numbers.
```
[the remark this answers is not in the document]
Comment (dadlerj) [reply client/jetbrains/webview/src/telemetry/EventLogger.ts:27]: Ugh this is a great point, and will require me to keep some scaffolding/a noop handler in place. Thanks for highlighting!

And to your question, this PR specifically removes that concern (and updates the admin dashboard to use the new data backend). It's not yet live.

[the remark this answers is not in the document]
Comment (dadlerj) [reply client/jetbrains/webview/src/telemetry/EventLogger.ts:27]: Actually I suppose I misinterpreted the problem here. I need to address a customer who has a modern IDE extension with an old server, and misses out on usage stats as a result...

Just to check, are IDE extensions auto-updated? Or is it possible that a customer would have a new server, with old IDE extensions?

File: internal/database/event_logs.go [hunk 2]
```diff
@@ -548,33 +558,36 @@ func (l *eventLogStore) countPerPeriodBySQL(ctx context.Context, countExpr, inte
...
 	}
-	return l.countUniqueUsersBySQL(ctx, startDate, endDate, sqlf.Sprintf("AND name IN (%s)", sqlf.Join(items, ",")))
+	return l.countUniqueUsersBySQL(ctx, startDate, endDate, []*sqlf.Query{sqlf.Sprintf("name IN (%s)", sqlf.Join(items, ","))})
 }
 
-func (l *eventLogStore) countUniqueUsersBySQL(ctx context.Context, startDate, endDate time.Time, querySuffix *sqlf.Query) (int, error) {
-	if querySuffix == nil {
-		querySuffix = sqlf.Sprintf("")
+func (l *eventLogStore) countUniqueUsersBySQL(ctx context.Context, startDate, endDate time.Time, conds []*sqlf.Query) (int, error) {
+	if conds == nil {
```
Comment (eseliger) [inline internal/database/event_logs.go:584]:
```suggestion
	if len(conds) == nil {
```

Comment (eseliger) [reply internal/database/event_logs.go:584]: I think otherwise a non-nil but 0-length slice could sneak in here and cause a SQL error.

**Review by eseliger — APPROVED** [review]
> One open question from my end: should I make it so that system users are excluded by default, instead of requiring the dev to set a flag?

No strong feelings about it. If you think it's risky if we forget it, I'd do it by default, otherwise really doesn't matter much. 

Read through the code, I must admit I don't know much about how all this legacy metrics stuff worked and if this is actually a replacement but I totally trust your screenshots and the code looks good! 
Tysm for doing this, really appreciated ✨

File: internal/database/event_logs.go (hunk 2 above)
Comment (unknwon) [reply internal/database/event_logs.go:584]: You meant `	if len(conds) == 0 {` right? 😁

Comment (eseliger) [reply internal/database/event_logs.go:584]: I did 🙃

Comment (unknwon) [reply internal/database/event_logs.go:584]: Updated your suggestion in-place.

Comment (dadlerj) [reply internal/database/event_logs.go]: Thanks! Just to confirm @unknwon , `len(nil)` is equal to `0` as well, right (so the nil check is still valid here)?

Comment (unknwon) [reply internal/database/event_logs.go]: Correct!

**Comment by DaedalusG** [comment]
Just tagging this guy in here: https://github.com/sourcegraph/sourcegraph/issues/22097 _Still observing these issues with some customers_

**Pull request merged**